include/swift/AST/DiagnosticsSema.def

```diff
@@ -1024,8 +1024,13 @@ NOTE(note_init_parameter,none,
 
 
 ERROR(missing_nullary_call,none,
-     "function produces expected type %0; did you mean to call it with '()'?",
-     (Type))
+     "expected an expression of type %0, but expression has function type "
+     "%1; did you mean to call the function with '()'?",
+     (Type, Type))
+ERROR(missing_nullary_call_decl,none,
+     "expected an expression of type %0, but %1 is a function "
+     "returning %0; did you mean to call it with '()'?",
```

Comment:
How about `cannot convert value of type %0 (reference to %1 %2) to expected type %3; did you mean to call %2 with '()'`? It seems that having diagnostic mention `expression` and `function` doesn't really add much useful information because these are both obvious from the type.

Reply:
Or if we choose to just mention the return type: `cannot convert reference to %1 %2 returning %0 to expected type %0; did you mean to call %2?`

Reply:
The problem is that it's not reference being converted but a type of the reference to something. I think it's okay to mention full type just to show that's it's a function with no arguments.

Reply:
I don't have a strong opinion, though a reference is still a value, or else there's the same issue with the canonical `cannot convert value of type ... to type` (but that's just how we phrase it outside the compiler).

> I think it's okay to mention full type just to show that's it's a function with no arguments

The dilemma about mentioning the full type is because nullary call syntax can also be used with default arguments. It might be appropriate to customize the message accordingly: either mention just the return type, mention that the function has all parameters defaulted, or both.